### PR TITLE
[mlir] Align num elements type to LLVM ArrayType

### PR DESCRIPTION
**Pull request by clementval**
MLIR LLMArrayType is using `unsigned` for the number of elements while LLVM ArrayType is using `uint64_t` https://github.com/llvm/llvm-project/blob/4ae896fe979b7db501cabde4b6b3504478958682/llvm/include/llvm/IR/DerivedTypes.h#L377

This leads to silent truncation when we use it for globals in flang. 

```
program test
  integer(8), parameter :: large = 2**30
  real,  dimension(large) :: bigarray
  common /c/ bigarray
  bigarray(999) = 666
end
```

The above program would result in a segfault since the global would be of size 0 because of the silent truncation. 

```
fir.global common @c_(dense<0> : vector<4294967296xi8>) : !fir.array<4294967296xi8>
```
became
```
llvm.mlir.global common @c_(dense<0> : vector<4294967296xi8>) {addr_space = 0 : i32} : !llvm.array<0 x i8>
```

This patch updates the definition of MLIR ArrayType to take `uint64_t` as argument of the number of elements to be compatible with LLVM.

### CODE REVIEW

**Comment by joker-eph** [comment]
I wished we'd use signed integer consistently to forbid wrap-around :(

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (clementval) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2403]: This is taking a very long time to translate so I'm wondering if we should use another form of attribute or if the translation is the problem. Any insight?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (joker-eph) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:635]: Why the change to std::vector?

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:635]: SmallVector limit is too small to hold this number of elements. Translation would break with an assertion.

Comment (joker-eph) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:635]: The comment in SmallVectorBase says:

```
/// The template parameter specifies the type which should be used to hold the
/// Size and Capacity of the SmallVector, so it can be adjusted.
/// Using 32 bit size is desirable to shrink the size of the SmallVector.
/// Using 64 bit size is desirable for cases like SmallVector<char>, where a
/// 32 bit size would limit the vector to ~4GB. SmallVectors are used for
/// buffering bitcode output - which can exceed 4GB.
```

And the size is picked with:

```
template <class T>
using SmallVectorSizeType =
    std::conditional_t<sizeof(T) < 4 && sizeof(void *) >= 8, uint64_t,
                       uint32_t>;
```

Unfortunately this isn't something that can be tweaked apparently :(


You're allocating 16GB of memory with this vector, no wonder the test is slow :)

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (joker-eph) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2403]: So an array of 4G constant (16GB  memory just for the array of pointers) is gonna kill a lot of machines out there :)

I don't have a suggestion here though right now. I think you can already that the type change with a simple round-trip, without testing the translation?

Comment (clementval) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2403]: Yeah unfortunately it is smth in fortran :(  I guess round trip will be enough.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (joker-eph) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:636]: We're in the splat case, we can have a fast path for zero by getting a `ConstantAggregateZero` instead here.

And if not, we can go the `ConstantDataArray` route instead when the element type is scalar.

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: 👍  Using `ConstantAggregateZero` improves things a lot. I updated the code to use it when constant is known to be zero. I reverted the std::vector to SmallVector since I don't really have a use case that would require it now.

Comment (joker-eph) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: A splat of non-zero would have the problem, wouldn't it?

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: I guess. I'll have a look at that as well.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Is there a efficient way to represent splat constant arrays with a non-zero value in LLVM?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:640]: Comment added. 

Do we want to have an if to switch between SmallVector and std::vector?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: That's a good question. This will still take long time to translate. 

```
llvm.mlir.global common @c_(dense<1> : vector<4294967296xi8>) {addr_space = 0 : i32} : !llvm.array<4294967296 x i8>
```

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:640]: Are you sure SmallVector has a size limitation that is different from `std::vector`? It has a fixed stack size but you can resize using a size_t value so I would expect 2^32 elements should be fine. I would drop the 8 that was there before that seems a bit arbitrary and just use `SmallVector<llvm::Constant *> constants(numElements, child);`.

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:640]: Let me try that.

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:640]: Size seems to be limited to `4294967295 `

```
SmallVector unable to grow. Requested capacity (4294967296) is larger than maximum value for size type (4294967295)
```

Comment (joker-eph) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:640]: I pasted the documentation of small vector in this PR already: https://github.com/llvm/llvm-project/pull/93230#discussion_r1612337183

Pasting here:

```
/// The template parameter specifies the type which should be used to hold the
/// Size and Capacity of the SmallVector, so it can be adjusted.
/// Using 32 bit size is desirable to shrink the size of the SmallVector.
/// Using 64 bit size is desirable for cases like SmallVector<char>, where a
/// 32 bit size would limit the vector to ~4GB. SmallVectors are used for
/// buffering bitcode output - which can exceed 4GB.
```

And the size is picked with:

```
template <class T>
using SmallVectorSizeType =
    std::conditional_t<sizeof(T) < 4 && sizeof(void *) >= 8, uint64_t,
                       uint32_t>;
```

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (joker-eph) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Did you see the second part of my comment when I opened this thread?

> And if not, we can go the ConstantDataArray route instead when the element type is scalar.

Have you looked at `ConstantDataArray` ? It won't be a splat, but still hopefully better?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:640]: Didn't see this comment. Let's use std:vector then!

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: As I can see we still need to pass a vector of the number of elements. Let me check if it improves anything.

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2403]:
```suggestion
// CHECK: @big_ = common global [4294967296 x i8] zeroinitializer
llvm.mlir.global common @big_(dense<0> : vector<4294967296xi8>) {addr_space = 0 : i32} : !llvm.array<4294967296 x i8>
```
nit: we usually put the check line above the ir line to match.

**Review by gysit — APPROVED** [review]
LGTM from my side modulo last comment.

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (clementval) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2403]: Ok I'll update that. In flang we have lot of tests with the check after ...

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (joker-eph) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: `ConstantArray` creation for scalar types will have to build a vector of the individual integers by extracting from the Constant object. So you're necessarily incurring more work in this case, see the code here for example: https://github.com/llvm/llvm-project/blob/53b904823cba4e93d58793b0804407f812477724/llvm/lib/IR/Constants.cpp#L1220-L1226

Now something a bit strange is that this is using a SmallVector as well, so it should fail the exact same way shouldn't it?

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Maybe the size limitation is not checked when using push_back?

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: I just tested it and it fails with the same error if you don't initialize the SmallVector and then push child to it in a loop.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Ah it pushes the integer to the `Elts` vector and not the Constant*. The integer is an i8 which is below 4 bytes so it uses a 64-bits size limit while for the Constant* the size is limited to 32-bits. I guess things would stop working if you use an i64 instead of an i8?

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: That makes sense.

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: @joker-eph Did you have time to have a look at the ConstantDataArray path?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (ftynse) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:639]:
```suggestion
          // TODO: Handle all compatible types. This code only handles i32.
```

**Review by ftynse — APPROVED** [review]

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:639]: Thanks for the review. I'm not sure we want to keep the `ConstantDataArray` path. I'm happy to add more types here if we want to keep it in this PR.

Comment (ftynse) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:639]: I suspect we want to have it for efficiency reasons. We have had compile time issues with large nested constants previously that I suspect using ConstantData avoids. This is perfectly fine for a separate PR.

Comment (clementval) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Ok! I have added at least all int types supported.

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2402]:
```suggestion
// CHECK: @big_ = common global [4294967296 x i8] zeroinitializer
```
ultra nit: missing space

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:637]: nit: can you drop the else since there is a return in the if body? That should save one indention for the else body.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:648]: nit: These else's are probably also redundant?

**Review by gysit — APPROVED** [review]
Still LGTM!

**Review by joker-eph — APPROVED** [review]
Thanks!

**Pull request merged**
